public/content/developers/tutorials/secret-state/index.md

```diff
@@ -0,0 +1,682 @@
+
+After reading this article you will know how to create this kind of secret state holding server, a client for showing the state, and an onchain component for communication between the two. The main tools we use will be:
+
+- [Zokrates](https://zokrates.github.io/) for zero-knowledge proofs and their verification.
+- [Typescript](https://www.typescriptlang.org/) for both the server and the client.
+- [Node](https://nodejs.org/en) for the server.
+- [Viem](https://viem.sh/) for communication with the Blockchain.
+- [MUD](https://mud.dev/) for onchain data management. Proper disclosure: While I'm writing this on my own time for Ethereum Foundation, my day job is working for [Lattice](https://lattice.xyz/) which makes this framework.
+- [React](https://react.dev/) for the user interface of the client.
+- [Vite](https://vitejs.dev/) for serving the client code.
+
+## Minesweeper example {#minesweeper}
+
+[Minesweeper](<https://en.wikipedia.org/wiki/Minesweeper_(video_game)>) is a game that includes a secret map with a minefield. The player chooses to dig in a specific location. If that location has a mine, it's game over. Otherwise, the player gets the number of mines in the eight squares surrounding that location.
+
+This application is written using [MUD](https://mud.dev/), a framework that lets us store data onchain using a [key-value database](https://aws.amazon.com/nosql/key-value/) and synchronize that data automatically with offchain components. In addition to synchronization, MUD makes it easy to provide access control, and for other users to [extend](https://mud.dev/guides/extending-a-world) our application permissionlessly.
+
+### Running the minesweeper example {#running-minesweeper-example}
+
+To run the minesweeper example:
+
+1. Make sure you [have the prerequisites installed](https://mud.dev/quickstart#prerequisites): [Node](https://mud.dev/quickstart#prerequisites), [Foundry](https://book.getfoundry.sh/getting-started/installation), [`git`](https://git-scm.com/downloads), and [`pnpm`](https://git-scm.com/downloads).
+
+2. Clone the repository.
+
+   ```sh copy
+   git clone https://github.com/qbzzt/20240901-secret-state.git
+   ```
+
+3. Install the packages.
+
+   ```sh copy
+   cd 20240901-secret-state/
+   pnpm install
+   ```
+
+4. Start the program (including an [anvil](https://book.getfoundry.sh/anvil/) blockchain) and wait.
+
+   ```sh copy
+   pnpm dev
+   ```
+
+   Note that the startup takes a long time. To see the progress, first use the down arrow to scroll to the _contracts_ tab to see the MUD contracts being deployed. When you get the message _Waiting for file changes…_, the contracts are deployed and further progress will happen in the _server_ tab. There, you wait until you get the message _Verifier address: 0x...._.
```

Comment:
I'm not getting this personally.. I get something like this, and then the process finishes.

```sh
> mud-template-react@ dev .
> mprocs
```

Nothing ends up being served to `localhost:3000` when I do this.

Reply:
I think it's because mprocs is not installed, at least for that version of node. I did two things that should fix it:

1. Removed the restriction on Node version
2. Suggested that people run `mprocs` directly rather than through `pnpm dev`.

Reply:
BTW, thank you. I'm really glad to make progress on this.